### PR TITLE
Honda Radarless: Create&send buttons to camera when engaged.

### PR DESCRIPTION
**Pull request by csouers**
**Description**: Fix two issues with Honda Bosch Radarless + stock ACC

---
Issue # 1: Intermittent failure to cancel/auto-resume.

Cause: The stock camera firmware in radarless models locks out button commands when button messages are received at the wrong time/with a bad counter. A number of consistently valid incoming frames must be received before the lockout is released. I didn't find any kind of feedback bits from the camera indicating a lockout.

Attempted fixes that were unreliable/ineffective:
- Spam quickly in between the car's frames with valid counters.
- Spam just after or before the car's frame is sent/forwarded with the proper counter as well as any other data bytes in the frame that OP doesn't normally send.
- Waiting a while and trying again after unsuccessful spamming (no counter matching on this). I didn't really like this anyway, so I quickly abandoned it.

Fix: Intercept the buttons frame at the panda when engaged then create & send to the camera at the normal rate - 25hz. Panda firmware forwards when disengaged. The cancel button is still spammed as before.

---
Issue # 2: Sudden ACC disengagement (Cruise is Off) after not touching the wheel for period of time.

Cause: When stock LKAS is enabled in the background, the stock wheel touch timer/nag expiring can disengage ACC. The alerts aren't forwarded by openpilot so the user is unaware of the issue. Users have also reported this issue saying that they never pushed the LKAS button, so there is a possibility that the car is pressing the LKAS button on its own for some reason. It doesn't happen consistently, so it's unclear what is really happening there.

Fixes: Send the LKAS button to turn off LKAS if LKAS_HUD from the camera shows that it is enabled. Also block pass-through of the LKAS button if pressed by the car/user when openpilot is enabled/intercepting to prevent the stock system from being turned on.

---

Broken route. Release3 : `a918d2895b3210a1|2022-08-15--12-43-11`

- 89.3: user engages ACC
- 90.2:  car sends LKAS button. stock LKAS turns on
- 102.2: user has been hands off too long. stock LKAS turns on the visual nag (not forwarded, so not seen)
- 112.2: longer still. stock LKAS now turns on the auditory nag  (not forwarded, so not heard)
- 118.1: after ~30 seconds since the user has last touched wheel (~86 seconds) and has also "ignored" the nag, ACC finally disengages

**Verification**: Tested by two users with 2022+ Honda Civic Sedan and also a Hatchback. One user has been driving on this fix for at least 9 months with no issues or recurrences of these two issues.

Good route from this PR : `a918d2895b3210a1|2024-01-27--10-03-34`

- 28.2 : user presses LKAS button. Panda is forwarding, so stock LKAS turns on.
- 30.0: user engages ACC.
- 30.4: openpilot engages and presses the LKAS button, which successfully turns off LKAS.
- 36.9: user presses LKAS button again as a test, which is successfully blocked by openpilot.

Another good route that's more recent: `a918d2895b3210a1|2024-02-23--19-10-17`

Prereqs:
- [ ] panda - https://github.com/commaai/panda/pull/1814
- [ ] dbc - https://github.com/commaai/opendbc/pull/998

Todo:
- [ ] test on new HR-V? Should be fine anyway as the sent button frames are otherwise typical.
- [x] test route
- [x] broken route example
- [ ] add updated civic2022 route to routes.py

### CODE REVIEW

**Comment by csouers** [comment]
This has some similar changes to https://github.com/commaai/openpilot/pull/30850, such as the dicts for the two button signals. @sunnyhaibin, let me know if you want attribution.

**Comment by Saturntime** [comment]
I’m having issue 2 I believe going to post route here in a few days

**Comment by vanillagorillaa** [comment]
Can confirm this fix works with my Acura Integra branch https://github.com/commaai/openpilot/pull/32056

**Comment by MarcoTheDingo** [comment]
Hey @csouers ! I occasionally have Issue 1. Do you still need a route, and if so, are you looking for something specific to test? I have a 2022 Civic Hatchback, Sport Touring Trim.

**Comment by MarcoTheDingo** [comment]
Hey @csouers it looks like the latest commit causes CAN errors and also the LKAS warning light lights on my dash. Routes:
- 5130484aa8069bad/00000093--d9777caa5a
- 5130484aa8069bad/00000094--0b89a3ca92

**Comment by sshane** [comment]
I see CI failures. Are you sure we can't improve the receptiveness of our button message without forwarding? I see we don't already set the counter to the one from the car, that could be a problem currently.

File: selfdrive/car/honda/hondacan.py [hunk 1]
```diff
@@ -167,7 +167,8 @@ def create_ui_commands(packer, CAN, CP, enabled, pcm_speed, hud, is_metric, acc_
     commands.append(packer.make_can_msg("ACC_HUD", CAN.pt, acc_hud_values))
 
   lkas_hud_values = {
-    'SET_ME_X41': 0x41,
+    'ENABLED': 1,
+    'SET_ME_X20': 0x20,
```
Comment (sshane) [inline selfdrive/car/honda/hondacan.py:171]: Does this maintain both MSB 30 and `LDW_OFF` at MSB 27?

File: selfdrive/car/honda/hondacan.py [hunk 2]
```diff
@@ -167,7 +167,8 @@ def create_ui_commands(packer, CAN, CP, enabled, pcm_speed, hud, is_metric, acc_
     commands.append(packer.make_can_msg("ACC_HUD", CAN.pt, acc_hud_values))
 
   lkas_hud_values = {
-    'SET_ME_X41': 0x41,
+    'ENABLED': 1,
```
Comment (sshane) [inline selfdrive/car/honda/hondacan.py:170]: Can you rename enabled to something more clear? We have LDW_ON/OFF in the same message

File: selfdrive/car/honda/hondacan.py (hunk 1 above)
Comment (csouers) [reply selfdrive/car/honda/hondacan.py:171]: These are still just for LKAS. I will rename `ENABLED`.

File: selfdrive/car/honda/hondacan.py (hunk 2 above)
Comment (csouers) [reply selfdrive/car/honda/hondacan.py:170]: Just enabled. Available might be a good alternative.

**Comment by csouers** [comment]
> I see CI failures. I would rather merge a change without a safety change first. Are you sure we can't improve the receptiveness of our button message without forwarding? I see we don't already set the counter to the one from the car, that could be a problem currently.

The camera fw seems to enforce a brief window for incoming frames. If the camera detects that something isn't being received as/when expected, it locks out button commands until the "issue" clears up for a brief period. 

As for what I've tried to make spamming better, my counter setting trials didn't improve spamming at all. Of the 4 or so methods I attempted, the two I can recall right now are:

1. Synchronizing to the car's button counter each time it was received by OP along with incrementing the counter.
2. Along with #1, I tested sending the spammed frames just ahead of when the car was expected to transmit its next frame. I hoped that the camera would accept our message instead of the car's, but it didn't. Intercepting when OP is engaged is the only thing that has been rock solid.

> Issue 2 is only an issue for stock ACC, correct? We should be able to stay engaged as we manage that with alpha openpilot long.

Yes, this entire PR only applies to stock ACC.

> We should split these two out into separate PRs, that should make things easier to review.

Will do.

**Comment by sshane** [comment]
We've moved the car interfacing code to our [`opendbc`](https://github.com/commaai/opendbc) repository, which is now the new home for car ports and fingerprints. Please re-open your pull request against [`opendbc`](https://github.com/commaai/opendbc) at your convenience by using this command below. This will transform all changes under `selfdrive/car/` to `opendbc_repo/opendbc/car/`. Make sure you have initialized submodules beforehand and have checked out the latest `opendbc` commit.

```bash
PR_NUMBER=33045
curl -L https://github.com/commaai/openpilot/pull/$PR_NUMBER.patch | sed -e 's/selfdrive\/car/opendbc_repo\/opendbc\/car/g' | git apply -v --reject
```

Simply replace the PR number with your own. Once done, add the files, fix any conflicts, and open a new PR. Alternatively, you may start a new PR from scratch if that is easier for you.

**Pull request closed without merging**